File: sample-coreapi/Sources/RTRViewController.m
// ABBYY® Mobile Capture © 2019 ABBYY Production LLC.
// ABBYY is a registered trademark or a trademark of ABBYY Software Ltd.

#import "RTRViewController.h"

#import <AVFoundation/AVFoundation.h>
#import <AbbyyRtrSDK/AbbyyRtrSDK.h>

typedef enum : NSInteger {
	RTRCoreAPIScenarioText,
	RTRCoreAPIScenarioBusinessCards,
	RTRCoreAPIScenarioCount,
} RTRCoreAPIScenarioType;

/// Cell ID for languagesTableView.
static NSString* const RTRTableCellID = @"RTRTableCellID";

@interface RTRViewController () <UIImagePickerControllerDelegate, UITableViewDelegate, UITableViewDataSource,
	UINavigationControllerDelegate>

/// Recognition languages table.
@property (nonatomic, weak) IBOutlet UITableView* languagesTableView;
/// Button for show / hide table with availabe Core API actions.
@property (nonatomic, weak) IBOutlet UIBarButtonItem* actionsButton;
/// Button for show / hide table with recognition languages.
@property (nonatomic, weak) IBOutlet UIBarButtonItem* recognitionLanguagesButton;

/// Open Photo Library button.
@property (nonatomic, weak) IBOutlet UIButton* selectImageButton;

/// Text view for recognized text.
@property (nonatomic, weak) IBOutlet UITextView* textView;
/// Recognition progress view.
@property (nonatomic, weak) IBOutlet UIProgressView* progressView;
/// Label for error or warning info.
@property (nonatomic, weak) IBOutlet UILabel* infoLabel;

/// Store selected image for re-recognition on languages changing.
@property (nonatomic, strong) UIImage* selectedImage;

@property (atomic, assign) NSUInteger lastTaskNumber;
@property (nonatomic, assign) RTRCoreAPIScenarioType currentCoreAPIScenario;
@property (nonatomic, assign) BOOL currentTableForLanguages;

@property (nonatomic, strong) NSArray<NSString*>* recognitionLanguages;

@end

#pragma mark -

@implementation RTRViewController {
	/// Engine for AbbyyRtrSDK.
	RTREngine* _engine;
	/// Selected recognition languages.
	NSMutableSet* _selectedRecognitionLanguages;
}

#pragma mark - UIView LifeCycle

- (void)viewDidLoad
{
	[super viewDidLoad];

	// Default recognition language.
	_selectedRecognitionLanguages = [[NSSet setWithObject:RTRLanguageNameEnglish] mutableCopy];

	[self.languagesTableView registerClass:[UITableViewCell class] forCellReuseIdentifier:RTRTableCellID];

	self.languagesTableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];

	self.languagesTableView.hidden = YES;
	[self.recognitionLanguagesButton setTitle:[self languagesButtonTitle]];

	self.textView.text = @"";
	[self.progressView setProgress:0];

	NSString* licensePath = [[[NSBundle mainBundle] bundlePath] stringByAppendingPathComponent:@"license"];
	_engine = [RTREngine sharedEngineWithLicenseData:[NSData dataWithContentsOfFile:licensePath]];
	if(_engine == nil) {
		self.selectImageButton.enabled = NO;
		[self updateLogMessage:@"Invalid License"];
		return;
	}

	_lastTaskNumber = 0;

	self.actionsButton.enabled = YES;
	self.recognitionLanguagesButton.enabled = YES;
}

- (BOOL)prefersStatusBarHidden
{
	return NO;
}

- (UIStatusBarStyle)preferredStatusBarStyle
{
	return UIStatusBarStyleLightContent;
}

#pragma mark - Actions

- (IBAction)capturePressed
{
	if(!self.selectImageButton.enabled) {
		return;
	}

	UIImagePickerController* photoLibrary = [[UIImagePickerController alloc] init];
	photoLibrary.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
	photoLibrary.delegate = self;
	photoLibrary.modalPresentationStyle = UIModalPresentationOverFullScreen;

	[self presentViewController:photoLibrary animated:YES completion:nil];
}

- (IBAction)onChangeAction
{
	self.currentTableForLanguages = NO;
	[self changeLanguagesTableVisibilty];
}

- (IBAction)onReconitionLanguages
{
	self.currentTableForLanguages = YES;
	[self changeLanguagesTableVisibilty];
}

- (void)changeLanguagesTableVisibilty
{
	if(self.languagesTableView.hidden) {
		[self.languagesTableView reloadData];
		[self showSettingsTable:YES];

	} else {
		[self tryToCloseLanguagesTable];
	}
}

- (void)tryToCloseLanguagesTable
{
	if(self.currentTableForLanguages) {
		if(_selectedRecognitionLanguages.count != 0) {
			[self showSettingsTable:NO];
		}
	} else {
		[self showSettingsTable:NO];
	}
}

#pragma mark -

- (NSString*)languagesButtonTitle
{
	if(_selectedRecognitionLanguages.count == 1) {
		return _selectedRecognitionLanguages.anyObject;
	}

	NSMutableString* resultTitle = [@"" mutableCopy];
	for(NSString* language in _selectedRecognitionLanguages) {
		if([language isEqualToString:RTRLanguageNameChineseSimplified]) {
			[resultTitle appendString:@"CHS "];
		} else if([language isEqualToString:RTRLanguageNameChineseTraditional]) {
			[resultTitle appendString:@"CHT "];
		} else {
			[resultTitle appendFormat:@"%@ ", [language substringToIndex:2].uppercaseString];
		}
	}

	return resultTitle;
}

#pragma mark -

- (NSArray*)recognitionLanguages
{
	if(_recognitionLanguages == nil) {
		_recognitionLanguages = @[
			RTRLanguageNameChineseSimplified,
			RTRLanguageNameChineseTraditional,
			RTRLanguageNameEnglish,
			RTRLanguageNameFrench,
			RTRLanguageNameGerman,
			RTRLanguageNameItalian,
			RTRLanguageNameJapanese,
			RTRLanguageNameKorean,
			RTRLanguageNamePolish,
			RTRLanguageNamePortugueseBrazilian,
			RTRLanguageNameRussian,
			RTRLanguageNameSpanish,
		];
	}
	return _recognitionLanguages;
}

- (void)updateLogMessage:(NSString*)message
{
	dispatch_async(dispatch_get_main_queue(), ^{
		self.infoLabel.text = message;
	});
}

- (void)showTextBlocks:(NSArray<RTRTextBlock*>*)textBlocks
{
	NSMutableString* text = [@"" mutableCopy];
	for(RTRTextBlock* block in textBlocks) {
		for(RTRTextLine* line in block.textLines) {
			[text appendFormat:@"%@\n", line.text];
		}
		[text appendString:@"\n"];
	}

	self.textView.text = text;
}

- (void)showDataFields:(NSArray<RTRDataField*>*)dataFields
{
	NSMutableString* text = [@"" mutableCopy];
	for(RTRDataField* field in dataFields) {
		[text appendFormat:@"%@: %@\n", field.name, field.text];
	}

	self.textView.text = text;
}

#pragma mark - UIImagePickerControllerDelegate

- (void)imagePickerController:(UIImagePickerController*)picker didFinishPickingMediaWithInfo:(NSDictionary*)info
{
	UIImage* image = [info valueForKey:UIImagePickerControllerOriginalImage];
	self.selectedImage = image;

	__weak typeof(self) weakSelf = self;
	typeof(^{}) completion = ^{
		[picker.presentingViewController dismissViewControllerAnimated:YES completion:^{
			[weakSelf recognizeImage:image];
		}];
	};

	if(NSThread.isMainThread) {
		completion();
	} else {
		dispatch_async(dispatch_get_main_queue(), completion);
	}
}

- (void)recognizeImage:(UIImage*)image
{
	self.textView.text = @"";
	self.infoLabel.text = @"";

	self.lastTaskNumber++;
	NSUInteger currentTaskNumber = self.lastTaskNumber;

	__weak typeof(self) weakSelf = self;
	RTRProgressCallbackBlock progressBlock = ^BOOL(NSInteger percentage, RTRCallbackWarningCode warningCode)
	{
		if(currentTaskNumber != weakSelf.lastTaskNumber) {
			return NO;
		}

		dispatch_async(dispatch_get_main_queue(), ^{
			[weakSelf.progressView setProgress:percentage * 1.f / 100 animated:YES];
			[weakSelf onWarning:warningCode];
		});

		return YES;
	};

	typeof(^{}) actionBlock = nil;

	id<RTRCoreAPI> coreAPI = [_engine createCoreAPI];
	[coreAPI.textRecognitionSettings setRecognitionLanguages:_selectedRecognitionLanguages];
	switch(self.currentCoreAPIScenario) {
		case RTRCoreAPIScenarioText:
		{
			actionBlock = ^{
				NSError* error;
				NSArray<RTRTextBlock*>* result = [coreAPI recognizeTextOnImage:image onProgress:progressBlock
					onTextOrientationDetected:nil error:&error];

				dispatch_async(dispatch_get_main_queue(), ^{
					if(currentTaskNumber == weakSelf.lastTaskNumber) {
						if(result == nil) {
							[weakSelf onError:error];
						} else {
							[weakSelf showTextBlocks:result];
						}
					}
					[weakSelf.progressView setProgress:1 animated:YES];
					[weakSelf.progressView setProgress:0 animated:NO];
				});
			};
			[coreAPI.textRecognitionSettings setRecognitionLanguages:_selectedRecognitionLanguages];
			break;
		}

		case RTRCoreAPIScenarioBusinessCards:
		{
			actionBlock = ^{
				NSError* error;
				NSArray<RTRDataField*>* result = [coreAPI extractDataFromImage:image onProgress:progressBlock
					onTextOrientationDetected:nil error:&error];

				dispatch_async(dispatch_get_main_queue(), ^{
					if(currentTaskNumber == weakSelf.lastTaskNumber) {
						if(result == nil) {
							[weakSelf onError:error];
						} else {
							[weakSelf showDataFields:result];
						}
					}
					[weakSelf.progressView setProgress:1 animated:YES];
					[weakSelf.progressView setProgress:0 animated:NO];
				});
			};
			[coreAPI.dataCaptureSettings.configureDataCaptureProfile setRecognitionLanguages:_selectedRecognitionLanguages];
			break;
		}
		default:
			break;
	}

	dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), actionBlock);
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController*)picker
{
	[picker.presentingViewController dismissViewControllerAnimated:YES completion:nil];
}

- (void)onWarning:(RTRCallbackWarningCode)warningCode
{
	NSString* message = [self stringFromWarningCode:warningCode];
	if(message.length > 0) {
		[self updateLogMessage:message];
	}
}

- (void)onError:(NSError*)error
{
	[self updateLogMessage:error.localizedDescription];
}

/// Human-readable descriptions for the RTRCallbackWarningCode constants.
- (NSString*)stringFromWarningCode:(RTRCallbackWarningCode)warningCode
{
	NSString* warningString;
	switch(warningCode) {
		case RTRCallbackWarningTextTooSmall:
			warningString = @"Text is too small.";
			break;

		case RTRCallbackWarningRecognitionIsSlow:
			warningString = @"The image is being recognized too slowly, perhaps something is going wrong.";
			break;

		case RTRCallbackWarningProbablyLowQualityImage:
			warningString = @"The image probably has low quality.";
			break;

		case RTRCallbackWarningProbablyWrongLanguage:
			warningString = @"The chosen recognition language is probably wrong.";
			break;

		case RTRCallbackWarningWrongLanguage:
			warningString = @"The chosen recognition language is wrong.";
			break;

		default:
			break;
	}

	return warningString;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView*)tableView didSelectRowAtIndexPath:(NSIndexPath*)indexPath
{
	if(self.currentTableForLanguages) {
		NSString* language = self.recognitionLanguages[indexPath.row];
		BOOL isSelected = ![_selectedRecognitionLanguages containsObject:language];
		if(isSelected) {
			[_selectedRecognitionLanguages addObject:language];
		} else {
			[_selectedRecognitionLanguages removeObject:language];
		}

		[self.recognitionLanguagesButton setTitle:[self languagesButtonTitle]];

		[tableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
	} else {
		self.currentCoreAPIScenario = indexPath.row;
		[self.actionsButton setTitle:[self titleForAction:self.currentCoreAPIScenario]];
		[self tryToCloseLanguagesTable];
	}
}

#pragma mark - UITableViewDatasource

- (NSInteger)tableView:(UITableView*)tableView numberOfRowsInSection:(NSInteger)section
{
	if(self.currentTableForLanguages) {
		return self.recognitionLanguages.count;
	} else {
		return RTRCoreAPIScenarioCount;
	}
}

- (UITableViewCell*)tableView:(UITableView*)tableView cellForRowAtIndexPath:(NSIndexPath*)indexPath
{
	UITableViewCell* cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:nil];
	if(self.currentTableForLanguages) {
		NSString* language = self.recognitionLanguages[indexPath.row];
		cell.textLabel.text = language;
		cell.accessoryType = [_selectedRecognitionLanguages containsObject:language]
			? UITableViewCellAccessoryCheckmark : UITableViewCellAccessoryNone;
	} else {
		cell.textLabel.text = [self titleForAction:indexPath.row];
		cell.accessoryType = self.currentCoreAPIScenario == indexPath.row
			? UITableViewCellAccessoryCheckmark : UITableViewCellAccessoryNone;
	}

	cell.backgroundColor = [UIColor colorWithRed:0 green:0 blue:0 alpha:0.4f];
	cell.textLabel.textColor = [UIColor whiteColor];
	cell.tintColor = [UIColor whiteColor];

	return cell;
}

- (NSString*)presetTitle:(NSString*)preset
{
	NSRange fstRange = [preset rangeOfCharacterFromSet:[NSCharacterSet decimalDigitCharacterSet]];
	if(fstRange.location == NSNotFound) {
		return preset;
	}

	return [preset substringFromIndex:fstRange.location];
}

#pragma mark - CoreAPI Actions

- (NSString*)titleForAction:(RTRCoreAPIScenarioType)action
{
	switch(action) {
		case RTRCoreAPIScenarioText:
			return @"Text";
			break;

		case RTRCoreAPIScenarioBusinessCards:
			return @"BusinessCards";
			break;

		default:
			break;
	}
	return nil;
}

#pragma mark - Utils

- (void)showSettingsTable:(BOOL)show
{
	self.languagesTableView.hidden = !show;
	[self updateLogMessage:nil];
	if(!show) {
		if(_selectedRecognitionLanguages.count == 0) {
			[_selectedRecognitionLanguages addObject:RTRLanguageNameEnglish];
			[self.recognitionLanguagesButton setTitle:[self languagesButtonTitle]];
		}
		if(self.selectedImage != nil) {
			[self recognizeImage:self.selectedImage];
		}
	}
}

@end

File: sample-coreapi-swift/Sources/RTRViewController.swift
/// ABBYY® Mobile Capture © 2019 ABBYY Production LLC.
/// ABBYY is a registered trademark or a trademark of ABBYY Software Ltd.


import AVFoundation
import AbbyyRtrSDK

// MARK: -

class RTRViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

	/// Open Photo Library button.
	@IBOutlet weak var selectImageButton: UIButton!

	@IBOutlet weak var actionsButton: UIBarButtonItem!
	/// Table with settings.
	@IBOutlet weak var tableView: UITableView!
	/// Button for show / hide table with settings.
	@IBOutlet weak var showSettingsButton: UIBarButtonItem!

	/// Label for error or warning info.
	@IBOutlet weak var infoLabel: UILabel!

	/// Text view for recognized text.
	@IBOutlet weak var textView: UITextView!
	/// Recognition progress view.
	@IBOutlet weak var progressView: UIProgressView!

	/// Stores selected image for re-recognition on languages changing.
	var selectedImage: UIImage!

	/// Engine for AbbyyRtrSDK.
	private var engine: RTREngine?
	/// Instance for core functionality.
	private var coreAPI: RTRCoreAPI?
	private var lastTaskNumber = 0
	/// Selected recognition languages.
	private var selectedRecognitionLanguages = Set([RTRLanguageName.english])

	/// Available recognition languages.
	private let recognitionLanguages: [RTRLanguageName] = [
		.chineseSimplified,
		.chineseTraditional,
		.english,
		.french,
		.german,
		.italian,
		.japanese,
		.korean,
		.polish,
		.portugueseBrazilian,
		.russian,
		.spanish
	]
	
	let avaliableScenarios = [
		"Text",
		"BusinessCards"
	]
	
	enum TableState {
		case scenarios
		case languages
		case none
	}
	
	private var currentCoreAPIScenario = "Text"
	private var currentTableState: TableState = .none {
		willSet {
			if currentTableState != .none {
				tryToCloseSettingsTable()
			}
		}
		didSet {
			DispatchQueue.main.async { [weak self] in
				self?.changeTableVisibilty()
			}
		}
	}
	
	private var tableContent: [String] = []

// MARK: - LifeCycle

	override func viewDidLoad()
	{
		super.viewDidLoad()

		actionsButton.isEnabled = true
		
		tableView.tableFooterView = UIView(frame: CGRect.zero)
		tableView.isHidden = true

		showSettingsButton.title = settingsButtonTitle()
		textView.text = ""

		progressView.setProgress(0, animated: false)

		let licensePath = (Bundle.main.bundlePath as NSString).appendingPathComponent("license")
		let licenseUrl = URL.init(fileURLWithPath: licensePath)
		if let data = try? Data(contentsOf: licenseUrl) {
			engine = RTREngine.sharedEngine(withLicense: data)
		}

		guard let _ = engine else {
			selectImageButton.isEnabled = false;
			updateLogMessage("Invalid License")
			return
		}
	}

	override var prefersStatusBarHidden: Bool
	{
		return true
	}

// MARK: - Private

	private func settingsButtonTitle() -> String
	{
		if self.selectedRecognitionLanguages.count == 1 {
			return self.selectedRecognitionLanguages.first!.rawValue
		}

		var languageCodes = [String]()

		for language in self.selectedRecognitionLanguages {
			let index = language.rawValue.index(language.rawValue.startIndex, offsetBy: 2)
			languageCodes.append(String(language.rawValue[..<index]))
		}

		return languageCodes.joined(separator: " ")
	}

	private func changeTableVisibilty()
	{
		switch currentTableState {
		case .scenarios:
			tableContent = avaliableScenarios
		case .languages:
			tableContent = recognitionLanguages.map{ $0.rawValue }
		default:
			tableContent = []
		}
		
		tableView.reloadData()
		tableView.isHidden = currentTableState == .none
	}

	private func tryToCloseSettingsTable()
	{
		if selectedRecognitionLanguages.isEmpty {
			selectedRecognitionLanguages.insert(.english)
		}

		showSettingsTable(show: false);
		showSettingsButton.title = settingsButtonTitle()

		if let image = selectedImage {
			recognizeImage(image)
		}
	}

	private func showSettingsTable(show: Bool)
	{
		tableView.isHidden = !show
		updateLogMessage("")
	}
	
	private func updateLogMessage(_ message: String?)
	{
		if let _message = message {
			infoLabel.text = _message
		} else {
			infoLabel.text = ""
		}
	}

// MARK: - UITableViewDelegate

	func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath)
	{
		switch currentTableState {
		case .languages:
			updateEnabledLanguages(with: RTRLanguageName(tableContent[indexPath.row]), indexPath: indexPath)
		case .scenarios:
			updateScenario(with: tableContent[indexPath.row])
		default:
			break
		}
	}
	
	private func updateEnabledLanguages(with language: RTRLanguageName, indexPath: IndexPath) {
		if !selectedRecognitionLanguages.contains(language) {
			selectedRecognitionLanguages.insert(language)
		} else {
			selectedRecognitionLanguages.remove(language)
		}
		showSettingsButton.title = settingsButtonTitle()
		tableView.reloadRows(at: [indexPath], with: .automatic)
	}
	
	private func updateScenario(with scenario: String) {
		currentCoreAPIScenario = scenario
		actionsButton.title = scenario
		tableView.reloadData()
		currentTableState = .none
	}

// MARK: - UITableViewDatasource

	func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int
	{
		return tableContent.count
	}

	func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell
	{
		let cellId = "cell id"
		var cell: UITableViewCell! = tableView.dequeueReusableCell(withIdentifier: cellId)
		if cell == nil {
			cell = UITableViewCell(style: UITableViewCell.CellStyle.subtitle, reuseIdentifier: cellId)
		}

		let title = tableContent[indexPath.row]
		cell.textLabel?.text = title
		
		switch currentTableState {
		case .languages:
			cell.accessoryType = selectedRecognitionLanguages.contains(RTRLanguageName(title)) ? .checkmark : .none
		case .scenarios:
			cell.accessoryType = title == currentCoreAPIScenario ? .checkmark : .none
		default:
			break
		}
		cell.backgroundColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0.4)
		cell.textLabel?.textColor = .white
		cell.tintColor = .white
		return cell
	}

// MARK: - Actions

	@IBAction func onSettingsButtonPressed()
	{
		if currentTableState == .languages {
			currentTableState = .none
		} else {
			currentTableState = .languages
		}
	}
	
	
	@IBAction func onChangeAction(_ sender: Any)
	{
		if currentTableState == .scenarios {
			currentTableState = .none
		} else {
			currentTableState = .scenarios
		}
	}
	
	@IBAction func onSelectImageButtonPressed()
	{
		if !selectImageButton.isEnabled {
			return
		}

		let imagePicker: UIImagePickerController = UIImagePickerController.init()
		imagePicker.sourceType = .savedPhotosAlbum
		imagePicker.delegate = self
		imagePicker.modalPresentationStyle = .fullScreen

		present(imagePicker, animated: true, completion: nil)
	}

// MARK: -

	func recognizeImage(_ image: UIImage)
	{
		textView.text = ""
		infoLabel.text = ""
		
		guard let engine = engine else { return }
		coreAPI = engine.createCoreAPI()
		lastTaskNumber += 1
		let currentTaskNumber = lastTaskNumber

		weak var weakSelf = self
		let progressBlock: (NSInteger, RTRCallbackWarningCode) -> Bool = { percentage, warningCode in
			guard currentTaskNumber == weakSelf?.lastTaskNumber else {
				return false
			}
			DispatchQueue.main.async {
				if let strongSelf = weakSelf {
					strongSelf.progressView.setProgress(Float(percentage) / 100.0, animated: true)
					strongSelf.onWarning(warningCode)
				}
			}
			return true;
		};
		
		let textActionBlock = {
			guard let coreAPI = weakSelf?.coreAPI else { return }
			do {
				let result = try coreAPI.recognizeText(on: image, onProgress: progressBlock, onTextOrientationDetected: nil)
				
				DispatchQueue.main.async {
					if let strongSelf = weakSelf, strongSelf.lastTaskNumber == currentTaskNumber {
						strongSelf.showTextBlocks(result);
						strongSelf.progressView.setProgress(1, animated: true)
						strongSelf.progressView.setProgress(0, animated: false)
					}
				}
			} catch let error as NSError {
				DispatchQueue.main.async {
					if let strongSelf = weakSelf {
						strongSelf.onError(error)
					}
				}
			}
		}
		
		let bcrActionBlock = {
			guard let coreAPI = weakSelf?.coreAPI else { return }
			do {
				let result = try coreAPI.extractData(from: image, onProgress: progressBlock, onTextOrientationDetected: nil)
				
				DispatchQueue.main.async {
					if let strongSelf = weakSelf, strongSelf.lastTaskNumber == currentTaskNumber {
						strongSelf.showDataFields(dataFields: result)
						strongSelf.progressView.setProgress(1, animated: true)
						strongSelf.progressView.setProgress(0, animated: false)
					}
					
				}
			} catch let error as NSError {
				DispatchQueue.main.async {
					if let strongSelf = weakSelf {
						strongSelf.onError(error)
					}
				}
			}
		}

		DispatchQueue.global(qos: .default).async {
			guard let strongSelf = weakSelf else { return }
			let selectedLanguages = strongSelf.selectedRecognitionLanguages
			switch strongSelf.currentCoreAPIScenario {
			case "Text":
				strongSelf.coreAPI?.textRecognitionSettings.setRecognitionLanguages(Set(selectedLanguages.map{ $0.rawValue }))
				textActionBlock()
			case "BusinessCards":
				_ = strongSelf.coreAPI?.dataCaptureSettings.configureDataCaptureProfile()?.setRecognitionLanguages(Set(selectedLanguages.map{ $0.rawValue }))
				bcrActionBlock()
			default:
				break
			}
		}
	}

	func onError(_ error: Error)
	{
		print(error.localizedDescription)

		var description = error.localizedDescription
		if description.contains("ChineseJapanese.rom") {
			description = "Chineze, Japanese and Korean are available in EXTENDED version only. Contact us for more information."
		} else if description.contains("KoreanSpecific.rom") {
			description = "Chineze, Japanese and Korean are available in EXTENDED version only. Contact us for more information."
		} else if description.contains("Russian.edc") {
			description = "Cyrillic script languages are available in EXTENDED version only. Contact us for more information."
		}

		updateLogMessage(description)
	}

	func onWarning(_ warningCode: RTRCallbackWarningCode)
	{
		let message = stringFromWarningCode(warningCode);
		updateLogMessage(message);
	}

	func showTextBlocks(_ textBlocks: [RTRTextBlock])
	{
		var text = ""
		for block in textBlocks {
			for textLine in block.textLines {
				text += textLine.text + "\n"
			}
		}
		textView.text = text
	}
	
	func showDataFields(dataFields: [RTRDataField])
	{
		textView.text = dataFields.map { "\($0.name ?? ""): \($0.text)" }.joined(separator: "\n")
	}

	/// Human-readable descriptions for the RTRCallbackWarningCode constants.
	func stringFromWarningCode(_ warningCode: RTRCallbackWarningCode) -> String
	{
		var warningString = ""
		switch warningCode {
		case .textTooSmall:
			warningString = "Text is too small."

		case .recognitionIsSlow:
			warningString = "The image is being recognized too slowly, perhaps something is going wrong."

		case .probablyLowQualityImage:
			warningString = "The image probably has low quality."

		case .probablyWrongLanguage:
			warningString = "The chosen recognition language is probably wrong."

		case .wrongLanguage:
			warningString = "The chosen recognition language is wrong."

		case .noWarning:
			break
		}
		return warningString
	}
}

// MARK: -

extension RTRViewController: UINavigationControllerDelegate
{
}

// MARK: -

extension RTRViewController: UIImagePickerControllerDelegate
{
	func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any])
	{
		if let image = info[UIImagePickerController.InfoKey.originalImage] {
			selectedImage = image as? UIImage

			weak var weakSelf = self

			let completion:() -> Void = {
				picker.presentingViewController?.dismiss(animated: true) {
					if let strongSelf = weakSelf {
						strongSelf.recognizeImage(strongSelf.selectedImage)
					}
				}
			}

			if(Thread.isMainThread) {
				completion();
			} else {
				DispatchQueue.main.async {
					completion()
				}
			}
		}
	}

	func imagePickerControllerDidCancel(_ picker: UIImagePickerController)
	{
		picker.presentingViewController?.dismiss(animated: true, completion: nil)
	}
}
